protobuf/src/main/scala/magnolify/protobuf/ProtobufType.scala

```diff
@@ -0,0 +1,171 @@
+}
+
+object ProtobufField {
+
+  trait Aux[T, From, To] extends ProtobufField[T] {
+    override type FromT = From
+    override type ToT = To
+  }
+
+  trait Record[T] extends Aux[T, Message, Message]
+
+  //////////////////////////////////////////////////
+
+  type Typeclass[T] = ProtobufField[T]
+
+  def combine[T](caseClass: CaseClass[Typeclass, T]): Record[T] = new Record[T] {
+
+    override def from(v: Message): T = {
+      caseClass.construct { p =>
+        val fieldDescriptor = v.getDescriptorForType.findFieldByName(p.label)
+
+        if (fieldDescriptor.getType == FieldDescriptor.Type.MESSAGE) {
+          // nested
+          p.typeclass.fromAny(v.getField(fieldDescriptor))
+        } else {
+          // non-nested
+          p.typeclass.fromAny(v.getField(fieldDescriptor))
+        }
+      }
+    }
+
+    override def to(v: T, bu: Message.Builder): Message = {
+
+      // clear builder from previous runs before using it to construct a new instance
+      caseClass.parameters
+        .foldLeft(bu.clear()) { (b, p) =>
+          val fieldDescriptor = b.getDescriptorForType.findFieldByName(p.label)
+
+          if (fieldDescriptor.getType == FieldDescriptor.Type.MESSAGE) { // nested records
+            val messageValue =
+              p.typeclass.to(p.dereference(v), b.newBuilderForField(fieldDescriptor))
+
+            if (messageValue == null) b else b.setField(fieldDescriptor, messageValue)
+          } else {
+            // non-nested
+            val fieldValue = p.typeclass.to(p.dereference(v), b)
+            if (fieldValue == null) b else b.setField(fieldDescriptor, fieldValue)
+          }
+        }
+        .build()
+    }
+  }
+
+  @implicitNotFound("Cannot derive ProtobufField for sealed trait")
+  private sealed trait Dispatchable[T]
+
+  def dispatch[T: Dispatchable](sealedTrait: SealedTrait[Typeclass, T]): Record[T] = ???
+
+  implicit def gen[T]: Record[T] = macro Magnolia.gen[T]
+
+  //////////////////////////////////////////////////
+
+  def apply[T](implicit f: ProtobufField[T]): ProtobufField[T] = f
+
+  //////////////////////////////////////////////////
+
+  private def at[T, From, To](f: From => T)(g: T => To): ProtobufField[T] =
+    new ProtobufField[T] {
+      override type FromT = From
+      override type ToT = To
+
+      override def from(v: FromT): T = f(v)
+
+      override def to(v: T, b: Message.Builder): ToT = g(v)
+    }
+
+  def from[T, Repr](f: Repr => T)(g: T => Repr): ProtobufField[T] =
+    at[T, Repr, Repr](f)(g)
+
+  private def id[T]: ProtobufField[T] =
+    at[T, T, T](identity)(identity)
+
+  implicit val pfBoolean = id[Boolean]
+  implicit val pfString = id[String]
+  implicit val pfInt = id[Int]
+  implicit val pfLong = id[Long]
+  implicit val pfFloat = id[Float]
+  implicit val pfDouble = id[Double]
+  implicit val pfBytes =
+    from[Array[Byte], ByteBuffer](bb => ju.Arrays.copyOfRange(bb.array(), bb.position(), bb.limit())
```

Comment:
Why do we need to support `ByteBuffer`? It's quite error-prone. `ByteString` is a safer interface with Protobuf.

Reply:
I copy/pasted this from somewhere, we don't have to support it. will update to ByteString :)